### ISSUE TITLE
Performance Improvements V2

### ISSUE THREAD

**Issue by mcMickJuice** [namespace-ee/react-calendar-timeline#338]
Let's discuss additional performance improvements that we can make to this library.  Here are a couple of my goals and also some that others have mentioned:

* lazy rendering - only render items that are in or near viewport.  This can be achieved by having an observer on each group row. If that group row is in or near viewport, then we render items in that group. If not, then we don't.  This has some drawbacks as you don't have ctrl + F functionality but we can probably get around that
* cut down on unnecessary renders - don't rerender on canvas mouse over when `showCursor` is true
* Utilize PureComponent

**Comment by jabidof** [comment]
Have a look at https://github.com/NoamELB/shouldComponentUpdate-Children/blob/master/README.md
Not the canonical approach I would like to see implemented (from an academically viewpoint) but I could get a performance boost of a factor of 6-8! 
For just a one-liner I adopted it 👏

**Comment by Ilaiwi** [comment]
an improvement would be to memorize the results from the `dimensionItems`, `groupHeights`, `groupTops` this will of course cut down extra renders as well...

**Comment by jabidof** [comment]
Any chance react waypoint would improve performances?
http://brigade.github.io/react-waypoint/

**Comment by mcMickJuice** [comment]
For sure @jabidof .  I was looking at that or using a polyfill for [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API).  Either way, I think my strategy for this would be:

* use two waypoints as a top and bottom "sentinel". These sentinels will be tied to a GroupRow so we will know what groups are within visible range (+ / - some padding)
* For those groups that are within visible range, only render Items tied to those groups (vertical culling)
* as you scroll, these waypoints will be hit and reset, which will update the range of visible groups

This is theoretical but that should cut down on rendering of items that are out of view.  The same approach can be applied horizontally.

**Comment by mcMickJuice** [comment]
This is a big one too: https://github.com/namespace-ee/react-calendar-timeline/issues/357

**Comment by acemac** [comment]
Would https://github.com/bvaughn/react-virtualized be possible for this?

Or any hints? I'm going to end up with thousands of groups and will need to cut down on re-render times for them.

**Comment by Ilaiwi** [comment]
Yes, it would. actually, I might experiment with it in the following two days

**Comment by Shvarts** [comment]
@Ilaiwi Did you have any success on experimenting with https://github.com/bvaughn/react-virtualized?

**Comment by Ilaiwi** [comment]
@Shvarts no yet. However, I just merged #392 which has a substantial performance boost with no 
API changes. I will release it by the end of the week. 
I will get to virtual lists sometime soon but I can't say when exactly since I am very busy with work.

**Comment by vasdee** [comment]
@Shvarts @Ilaiwi shameless plug, but the timeline component mentioned under https://github.com/bvaughn/react-virtualized#friends was heavily inspired by this timeline and got some great performance gains out of react-virtualised. Full disclaimer, I'm a co-author

**Comment by hellooperatorpatchmethrough** [comment]
Any status updates ?

**Comment by Ilaiwi** [comment]
@hellooperatorpatchmethrough I am currently working on an RFC for `rowRenderer` which should change how the items are being redered and will improve the performance of the library a lot!

**Comment by developer-ocansey** [comment]
Closing as part of issue triage. This issue was created before 2023 and has had no recent activity. If you're still experiencing this on v0.30+, please open a new issue with a minimal reproduction.
